experimental/GraphicalModels/src/PhylogeneticModels.jl

```diff
@@ -0,0 +1,271 @@
+    
+    group = [[0,0], [0,1], [1,0], [1,1]]
+
+    return PhylogeneticModel(ns, graph, R, S, matrices, fourier_param, group)
+end
+
+# general Markov model
+# work in progress
+# not sure whether to use this https://docs.sciml.ai/Symbolics/stable/manual/arrays/
+# or array of polynomial ring elements in OSCAR
+function generalmarkov_model(graph::Graph{Directed})
+    ns = 4
+    ne = n_edges(graph)
+    R, list_m = polynomial_ring(QQ, :m => 1:(ne^2))
+    
+    matrices = Dict{Edge, MatElem}(e => matrix(R, [
+      m11 m12 m13 m14
+      m11 m12 m13 m14
+      b b a b
+      b b b a]) for (a,b,e) in zip(list_a, list_b, edges(graph))
+    )
+    return PhylogeneticModel(ns, graph, R, matrices)
+end
+
+# more models to come, Ch Mar 21
+
+
+#### SPECIALIZED FOURIER TRANSFORM MATRIX ####
+
+function hadamardmatrix()
+  H = [1 1 1 1
+       1 1 -1 -1 
+       1 -1 1 -1
+       1 -1 -1 1]
+  return H
+end
+
+# TODO: line 138 mutates f_equivclasses, i.e. an object from outside the function.
+# This makes the function a !-function. We do not want f_equivclasses to be changed, 
+# so I add the missing equivalence class again in line 175. 
+# We should find out whether there is a better way to do this.
+function specialized_fourier_transform(pm::PhylogeneticModel, p_equivclasses::Dict{Vector{Vector{Int64}}, QQMPolyRingElem},f_equivclasses::Dict{Vector{Vector{Int64}}, QQMPolyRingElem})
+  R = probabilities_ring(pm)
+
+  class0 = findall(x -> x ==0, f_equivclasses)[1]
+  delete!(f_equivclasses, class0)
+
+  np = length(p_equivclasses)
+  nq = length(f_equivclasses)
+
+  ## We need to sort the equivalence classes: both inside each class as well as the collection of classes. 
+  p_equivclasses_sorted = collect(keys(p_equivclasses))
+  for p_eqclass in p_equivclasses_sorted
+      sort!(p_eqclass)
+  end
+  sort!(p_equivclasses_sorted)
+
+  f_equivclasses_sorted = collect(keys(f_equivclasses))
+  for f_eqclass in f_equivclasses_sorted
+      sort!(f_eqclass)
+  end
+  sort!(f_equivclasses_sorted)
+
+  H = hadamardmatrix()
+
+  specialized_ft_matrix = R.(Int.(zeros(nq, np)))
+  for i in 1:nq
+      current_fourier_classes = f_equivclasses_sorted[i]
+      for j in 1:np
+          current_prob_classes = p_equivclasses_sorted[j]
+          current_entriesin_M = [prod([H[y,x] for (x,y) in zip(p,q)]) for p in current_prob_classes, q in current_fourier_classes]
+          specialized_ft_matrix[i,j] = R.(1//(length(current_prob_classes)*length(current_fourier_classes))*sum(current_entriesin_M))
+      end
+  end
+  get!(f_equivclasses, class0, R(0))
+  return specialized_ft_matrix
+end
+
+function inverse_specialized_fourier_transform(pm::PhylogeneticModel, p_equivclasses::Dict{Vector{Vector{Int64}}, QQMPolyRingElem},f_equivclasses::Dict{Vector{Vector{Int64}}, QQMPolyRingElem})
+  R = probabilities_ring(pm)
+
+  class0 = findall(x -> x ==0, f_equivclasses)[1]
+  delete!(f_equivclasses, class0)
```

Comment:
Maybe this can be simplified to this?
```suggestion
  filter!(!is_zero, f_equivclasses)
```